### PR TITLE
Add component alias into runtime machine if present in origin devfile

### PR DESCRIPTION
**Pull request by mshaposhnik**
# What does this PR do?
Adds  component alias into runtime machine if thay are present in origin devfile

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/14280

### CODE REVIEW

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/DockerimageComponentToWorkspaceApplier.java [hunk 1]
```diff
@@ -134,6 +135,10 @@ public void apply(
           .put(PROJECTS_VOLUME_NAME, new VolumeImpl().withPath(projectFolderPath));
     }
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/DockerimageComponentToWorkspaceApplier.java:137]: Adding alias attribute for dockerimage

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java [hunk 2]
```diff
@@ -144,15 +149,22 @@ public void apply(
...
+
+    linkCommandsToMachineName(workspaceConfig, k8sComponent, podToContainerAndMachineName);
 
     if (Boolean.TRUE.equals(k8sComponent.getMountSources())) {
       applyProjectsVolumes(podsData, componentObjects);
     }
 
     applyEntrypoints(k8sComponent.getEntrypoints(), componentObjects);
 
-    k8sEnvProvisioner.provision(workspaceConfig, environmentType, componentObjects, emptyMap());
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java:155]: Adding alias attribute for K8s/OS by pre-creating machine configs with given attribute.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java [hunk 3]
```diff
@@ -68,12 +77,40 @@ public MachineResolver(
...
-            null, toServers(containerEndpoints), null, null, toWorkspaceVolumes(cheContainer));
+            null,
+            toServers(containerEndpoints),
+            null,
+            toMachineAttributes(pluginId, wsAttributes),
+            toWorkspaceVolumes(cheContainer));
 
     normalizeMemory(container, machineConfig);
     return machineConfig;
   }
```
Comment (mshaposhnik) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java:89]: Adding alias attribute for cheEditor/Plugin by checking for `pluginID=alias` pair in appropriate WS config attributes;

**Review by skabashnyuk — COMMENTED** [review]
in general look ok to me. @sleshchenko @metlos @sparkoo wdyt?

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java [hunk 4]
```diff
@@ -144,15 +149,22 @@ public void apply(
 
     List<PodData> podsData = getPodDatas(componentObjects);
 
-    estimateCommandsMachineName(workspaceConfig, k8sComponent, podsData);
+    Map<String, Map<String, String>> podToContainerAndMachineName = estimateMachineNames(podsData);
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: I don't like this `Map<String, Map<String, String>>`, it's very hard to follow. Does it really have to be map of maps ? I can't find any usage of it's keys.

**Review by sparkoo — COMMENTED** [review]
General logic seems ok. But it looks like implementation could be simplified. Many string maps and map of maps smells...

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java (hunk 4 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: It's because containers in different pods may have same names. So need to store them separately. I will try to find way to simplify, but not sure if i could.

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/DockerimageComponentToWorkspaceApplier.java (hunk 1 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/DockerimageComponentToWorkspaceApplier.java:137]: there is check

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java [hunk 5]
```diff
@@ -192,6 +204,41 @@ private void applyProjectsVolumes(List<PodData> podsData, List<HasMetadata> comp
     }
   }
 
+  /**
+   * Creates map of machine names and corresponding {@link MachineConfigImpl} with component alias
+   * attribute set.
+   */
+  private Map<String, MachineConfigImpl> prepareMachineConfigs(
+      String componentAlias, Map<String, Map<String, String>> machineNamesData) {
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: ok

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java [hunk 6]
```diff
@@ -235,8 +282,10 @@ private String retrieveContent(Component recipeComponent, FileContentProvider fi
    *
    * <p>Machine name will be set only if the specified recipe objects has the only one container.
    */
-  private void estimateCommandsMachineName(
-      WorkspaceConfig workspaceConfig, Component component, List<PodData> podsData) {
+  private void linkCommandsToMachineName(
+      WorkspaceConfig workspaceConfig,
+      Component component,
+      Map<String, Map<String, String>> namesData) {
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: ok

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java (hunk 4 above)
Comment (sleshchenko) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: I can propose to go in another direction. You don't really know pods and containers during commands machine name estimation, so it's what I tried to implement (but not tested):
```java
public void apply ...
   List<PodData> podsData = getPodDatas(componentObjects);

    if (Boolean.TRUE.equals(k8sComponent.getMountSources())) {
      applyProjectsVolumes(podsData, componentObjects);
    }

    applyEntrypoints(k8sComponent.getEntrypoints(), componentObjects);

    Map<String, MachineConfigImpl> machineConfigs =
        k8sComponent.getAlias() != null
            ? prepareMachineConfigs(podsData, k8sComponent.getAlias())
            : emptyMap();

    linkCommandsToMachineName(workspaceConfig, k8sComponent, machineConfigs.keySet());
  ...

  /**
   * Creates map of machine names and corresponding {@link MachineConfigImpl} with component alias
   * attribute set.
   */
  private Map<String, MachineConfigImpl> prepareMachineConfigs(List<PodData> podsData,
      @Nullable String componentAlias) {
    Map<String, MachineConfigImpl> machineConfigs = new HashMap<>();
    for (PodData podData : podsData) {
      for (Container container : podData.getSpec().getContainers()) {
        String machineName = machineName(podData, container);

        MachineConfigImpl config = new MachineConfigImpl();
        if (Strings.isNullOrEmpty(componentAlias)) {
          config.getAttributes().put(DEVFILE_COMPONENT_ALIAS_ATTRIBUTE, componentAlias);
        }

        machineConfigs.put(machineName, config);
      }
    }
    return machineConfigs;
  }

/**
   * Set {@link Command#MACHINE_NAME_ATTRIBUTE} to commands which are configured in the specified
   * component.
   *
   * <p>Machine name will be set only if the specified recipe objects has the only one container.
   */
  private void linkCommandsToMachineName(
      WorkspaceConfig workspaceConfig,
      Component component,
      Set<String> machinesNames) {
    List<? extends Command> componentCommands =
        workspaceConfig
            .getCommands()
            .stream()
            .filter(
                c ->
                    component.getAlias() != null
                        && component
                            .getAlias()
                            .equals(
                                c.getAttributes().get(Constants.COMPONENT_ALIAS_COMMAND_ATTRIBUTE)))
            .collect(toList());
    if (componentCommands.isEmpty()) {
      return;
    }

    if (machinesNames.size() != 1) {
      // many or no pods - can't estimate the name because of ambiguity or lack of information
      return;
    }

    String machineName = machinesNames.iterator().next();
    componentCommands.forEach(c -> c.getAttributes().put(MACHINE_NAME_ATTRIBUTE, machineName));
  }
```

```

Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/devfile/KubernetesComponentToWorkspaceApplier.java]: Ok i will apply the approach with some minot edits, thanks

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java [hunk 7]
```diff
@@ -68,12 +77,40 @@ public MachineResolver(
...
 
+  private Map<String, String> toMachineAttributes(
```
Comment (sleshchenko) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java:90]: I think it would be better to rewrite it in more `Single Responsibility Principle` style

```java

  private Map<String, String> toMachineAttributes(String pluginId, Map<String, String> wsAttributes) {
   Map<String, String> attributes = new HashMap<>();

    Optional<String> pluginAlias = findPluginAlias(pluginId, wsAttributes);
    pluginAlias.ifPresent(s -> attributes.put(DEVFILE_COMPONENT_ALIAS_ATTRIBUTE, s));

    return attributes;
  }

  private Optional<String> findPluginAlias(String pluginId, Map<String, String> wsAttributes) {

    List<String> aliases = new ArrayList<>();

    String pluginComponentAliases = wsAttributes.get(PLUGINS_COMPONENTS_ALIASES_WORKSPACE_ATTRIBUTE);
    if (!isNullOrEmpty(pluginComponentAliases)) {
      aliases.addAll(asList(pluginComponentAliases.split(",")));
    }

    String editorComponentAlias = wsAttributes.get(EDITOR_COMPONENT_ALIAS_WORKSPACE_ATTRIBUTE);
    if (!isNullOrEmpty(editorComponentAlias)) {
      aliases.add(editorComponentAlias);
    }

    if (aliases.isEmpty()) {
      return Optional.empty();
    }

    return aliases
          .stream()
          .map(value -> value.split("="))
          .filter(arr -> arr[0].equals(pluginId))
          .map(arr -> arr[1])
          .findAny();
  }

```

**Review by sleshchenko — APPROVED** [review]
Good work :+1:

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java (hunk 7 above)
Comment (mshaposhnik) [reply infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/wsplugins/MachineResolver.java:90]: makes sense

**Comment by mshaposhnik** [comment]
crw-ci-test

**Comment by mshaposhnik** [comment]
ci-test

**Comment by mshaposhnik** [comment]
crw-ci-test

**Comment by mshaposhnik** [comment]
ci-test

**Comment by mshaposhnik** [comment]
crw-ci-test

**Review by metlos — APPROVED** [review]
